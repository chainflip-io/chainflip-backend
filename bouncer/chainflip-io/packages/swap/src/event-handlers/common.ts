import { z } from 'zod';
import { assetChains } from '@/shared/enums';
import {
  btcAddress,
  chainflipChain,
  dotAddress,
  hexString,
  unsignedInteger,
} from '@/shared/parsers';

export const egressId = z.tuple([
  z.object({ __kind: chainflipChain }).transform(({ __kind }) => __kind),
  unsignedInteger,
]);

const ethChainAddress = z.object({
  __kind: z.literal('Eth'),
  value: hexString,
});
const dotChainAddress = z.object({
  __kind: z.literal('Dot'),
  value: dotAddress,
});
const btcChainAddress = z.object({
  __kind: z.literal('Btc'),
  value: hexString
    .transform((v) => Buffer.from(v.slice(2), 'hex').toString())
    .pipe(btcAddress),
});

export const encodedAddress = z
  .union([ethChainAddress, dotChainAddress, btcChainAddress])
  .transform(
    ({ __kind, value }) =>
      ({
        chain: assetChains[__kind.toUpperCase() as Uppercase<typeof __kind>],
        address: value,
      } as const),
  );
