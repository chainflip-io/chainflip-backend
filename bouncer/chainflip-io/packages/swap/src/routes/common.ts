import { RequestHandler } from 'express';
import type { RouteParameters, Response } from 'express-serve-static-core';
import type { ParsedQs } from 'qs';
import logger from '../utils/logger';
import ServiceError from '../utils/ServiceError';

export const handleError = (res: Response, error: unknown) => {
  logger.customInfo('received error', {}, { error });

  if (error instanceof ServiceError) {
    res.status(error.code).json({ message: error.message });
  } else {
    logger.customError(
      'unknown error occurred',
      { alertCode: 'UnknownError' },
      { error },
    );
    res.status(500).json({ message: 'unknown error' });
  }
};

/* eslint-disable @typescript-eslint/no-explicit-any */
export const asyncHandler = <
  Route extends string,
  P = RouteParameters<Route>,
  ResBody = any,
  ReqBody = any,
  ReqQuery = ParsedQs,
  Locals extends Record<string, any> = Record<string, any>,
>(
  handler: RequestHandler<P, ResBody, ReqBody, ReqQuery, Locals>,
): typeof handler =>
  (async (req, res, next) => {
    try {
      await handler(req, res, next);
    } catch (error) {
      handleError(res, error);
    }
  }) as RequestHandler<P, ResBody, ReqBody, ReqQuery, Locals>;
/* eslint-enable @typescript-eslint/no-explicit-any */
