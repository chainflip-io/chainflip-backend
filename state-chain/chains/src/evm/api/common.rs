use super::*;
use crate::eth::deposit_address::get_salt;
use cf_primitives::{AssetAmount, ChannelId};
use ethabi::{Address, ParamType, Token, Uint};

#[derive(Encode, Decode, TypeInfo, Clone, RuntimeDebug, Default, PartialEq, Eq)]
pub(crate) struct EncodableFetchAssetParams {
	pub contract_address: Address,
	pub asset: Address,
}

impl Tokenizable for EncodableFetchAssetParams {
	fn tokenize(self) -> Token {
		Token::Tuple(vec![Token::Address(self.contract_address), Token::Address(self.asset)])
	}

	fn param_type() -> ethabi::ParamType {
		ParamType::Tuple(vec![ParamType::Address, ParamType::Address])
	}
}

#[derive(Encode, Decode, TypeInfo, Clone, RuntimeDebug, Default, PartialEq, Eq)]
pub(crate) struct EncodableFetchDeployAssetParams {
	pub channel_id: ChannelId,
	pub asset: Address,
}

impl Tokenizable for EncodableFetchDeployAssetParams {
	fn tokenize(self) -> Token {
		Token::Tuple(vec![
			Token::FixedBytes(get_salt(self.channel_id).to_vec()),
			Token::Address(self.asset),
		])
	}

	fn param_type() -> ethabi::ParamType {
		ParamType::Tuple(vec![ParamType::FixedBytes(32), ParamType::Address])
	}
}

#[derive(Encode, Decode, TypeInfo, Clone, RuntimeDebug, Default, PartialEq, Eq)]
pub(crate) struct EncodableTransferAssetParams {
	/// For EVM, the asset is encoded as a contract address.
	pub asset: Address,
	pub to: Address,
	pub amount: AssetAmount,
}

impl Tokenizable for EncodableTransferAssetParams {
	fn tokenize(self) -> Token {
		Token::Tuple(vec![
			Token::Address(self.asset),
			Token::Address(self.to),
			Token::Uint(Uint::from(self.amount)),
		])
	}

	fn param_type() -> ethabi::ParamType {
		ParamType::Tuple(vec![ParamType::Address, ParamType::Address, ParamType::Uint(256)])
	}
}
