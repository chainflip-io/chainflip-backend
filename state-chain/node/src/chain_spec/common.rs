use cf_primitives::AuthorityCount;
use sp_runtime::{Percent, Permill};
pub use state_chain_runtime::constants::common::*;
use state_chain_runtime::{chainflip::Offence, BlockNumber, FlipBalance, SetSizeParameters};

pub const GENESIS_FUNDING_AMOUNT: FlipBalance = 5_000 * FLIPPERINOS_PER_FLIP;
pub const MIN_FUNDING: FlipBalance = 10 * FLIPPERINOS_PER_FLIP;
pub const REDEMPTION_TAX: FlipBalance = 5 * FLIPPERINOS_PER_FLIP;
pub const MIN_AUTHORITIES: AuthorityCount = 2;
pub const AUCTION_PARAMETERS: SetSizeParameters = SetSizeParameters {
	min_size: MIN_AUTHORITIES,
	max_size: MAX_AUTHORITIES,
	max_expansion: MAX_AUTHORITIES,
};

/// Percent of the epoch we are allowed to redeem
pub const REDEMPTION_PERIOD_AS_PERCENTAGE: u8 = 50;

// Consider the equation (1 + x/1_000_000_000)^n = 1 + inf/100
// where inf is the target yearly inflation (percent), n is the number of compundings that
// we do in a year and x is the inflation rate (Perbill) for each compunding time period.
//
// The following values are calculated by solving the above equation for x using n =
// (365*14400)/150 (since compunding is done every heartbeat which is every 150 blocks) and
// inf is taken as 0.1 percent for authority emissions and 0.02 percent for backup node emissions.
//
// Can be generated using the following python code:
//
// ```
// import math
// def per_bill_inflation(pct):
//      return (math.pow(1 + pct/100, 1/35040) - 1) * 1000000000
//
// # Testnet:
// round(per_bill_inflation(0.1)) -> 28
// round(per_bill_inflation(0.02)) -> 6
// # Mainnet:
// round(per_bill_inflation(7)) -> 1931
// round(per_bill_inflation(1)) -> 284
// ```
//
pub const CURRENT_AUTHORITY_EMISSION_INFLATION_PERBILL: u32 = 28;
pub const BACKUP_NODE_EMISSION_INFLATION_PERBILL: u32 = 6;

pub const SUPPLY_UPDATE_INTERVAL: u32 = 24 * HOURS;

// This is equivalent to one reputation point for every minute of online time.
pub const REPUTATION_PER_HEARTBEAT: i32 = 15;
pub const ACCRUAL_RATIO: (i32, u32) = (REPUTATION_PER_HEARTBEAT, HEARTBEAT_BLOCK_INTERVAL);

const REPUTATION_PENALTY_SMALL: i32 = REPUTATION_PER_HEARTBEAT; // 15 minutes to recover reputation
const REPUTATION_PENALTY_MEDIUM: i32 = REPUTATION_PER_HEARTBEAT * 4; // One hour to recover reputation
const REPUTATION_PENALTY_LARGE: i32 = REPUTATION_PER_HEARTBEAT * 8; // Two hours to recover reputation

/// The offences committable within the protocol and their respective reputation penalty and
/// suspension durations.
pub const PENALTIES: &[(Offence, (i32, BlockNumber))] = &[
	(Offence::MissedHeartbeat, (REPUTATION_PENALTY_SMALL, 0)),
	(Offence::ParticipateKeygenFailed, (REPUTATION_PENALTY_MEDIUM, HEARTBEAT_BLOCK_INTERVAL)),
	(Offence::ParticipateSigningFailed, (REPUTATION_PENALTY_MEDIUM, MINUTES / 2)),
	(Offence::MissedAuthorshipSlot, (REPUTATION_PENALTY_LARGE, HEARTBEAT_BLOCK_INTERVAL)),
	(Offence::FailedToBroadcastTransaction, (REPUTATION_PENALTY_MEDIUM, HEARTBEAT_BLOCK_INTERVAL)),
	(Offence::GrandpaEquivocation, (REPUTATION_PENALTY_LARGE, HEARTBEAT_BLOCK_INTERVAL * 5)),
];

/// Daily slashing rate 0.1% (of the bond) for offline authority
pub const DAILY_SLASHING_RATE: Permill = Permill::from_perthousand(1);

/// Redemption delay on testnets is 2 MINUTES.
/// We use a ttl of 1 hour to give enough of a buffer.
pub const REDEMPTION_TTL_SECS: u64 = 2 * 3600;

/// Determines the expiry duration for governance proposals.
pub const EXPIRY_SPAN_IN_SECONDS: u64 = 24 * 3600;

pub const AUCTION_BID_CUTOFF_PERCENTAGE: Percent = Percent::from_percent(10);
