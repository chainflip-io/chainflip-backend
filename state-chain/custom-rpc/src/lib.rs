use cf_amm::common::SqrtPriceQ64F96;
use cf_chains::{btc::BitcoinNetwork, dot::PolkadotHash, eth::api::EthereumChainId};
use cf_primitives::{Asset, EthereumAddress, SwapOutput};
use jsonrpsee::{core::RpcResult, proc_macros::rpc, types::error::CallError};
use pallet_cf_governance::GovCallHash;
use sc_client_api::HeaderBackend;
use serde::{Deserialize, Serialize};
use sp_api::BlockT;
use sp_rpc::number::NumberOrHex;
use sp_runtime::AccountId32;
use state_chain_runtime::{
	chainflip::Offence,
	constants::common::TX_FEE_MULTIPLIER,
	runtime_apis::{ChainflipAccountStateWithPassive, CustomRuntimeApi, Environment},
};
use std::{marker::PhantomData, sync::Arc};

#[derive(Serialize, Deserialize)]
pub struct RpcAccountInfo {
	pub balance: NumberOrHex,
	pub bond: NumberOrHex,
	pub last_heartbeat: u32,
	pub is_live: bool,
	pub is_activated: bool,
	pub online_credits: u32,
	pub reputation_points: i32,
	pub state: ChainflipAccountStateWithPassive,
}

#[derive(Serialize, Deserialize)]
pub struct RpcAccountInfoV2 {
	pub balance: NumberOrHex,
	pub bond: NumberOrHex,
	pub last_heartbeat: u32,
	pub online_credits: u32,
	pub reputation_points: i32,
	pub keyholder_epochs: Vec<u32>,
	pub is_current_authority: bool,
	pub is_current_backup: bool,
	pub is_qualified: bool,
	pub is_online: bool,
	pub is_bidding: bool,
	pub bound_redeem_address: Option<EthereumAddress>,
}

#[derive(Serialize, Deserialize)]
pub struct RpcPenalty {
	reputation_points: i32,
	suspension_duration_blocks: u32,
}

type RpcSuspensions = Vec<(Offence, Vec<(u32, AccountId32)>)>;

#[derive(Serialize, Deserialize)]
pub struct RpcAuctionState {
	blocks_per_epoch: u32,
	current_epoch_started_at: u32,
	redemption_period_as_percentage: u8,
	min_funding: NumberOrHex,
	auction_size_range: (u32, u32),
}

#[derive(Serialize, Deserialize)]
pub struct RpcSwapOutput {
	// Intermediary amount, if there's any
	pub intermediary: Option<NumberOrHex>,
	// Final output of the swap
	pub output: NumberOrHex,
}

impl From<SwapOutput> for RpcSwapOutput {
	fn from(swap_output: SwapOutput) -> Self {
		Self {
			intermediary: swap_output.intermediary.map(NumberOrHex::from),
			output: NumberOrHex::from(swap_output.output),
		}
	}
}

#[derive(Serialize, Deserialize)]
pub struct RpcEnvironment {
	bitcoin_network: BitcoinNetwork,
	ethereum_chain_id: EthereumChainId,
	polkadot_genesis_hash: PolkadotHash,
}

impl From<Environment> for RpcEnvironment {
	fn from(environment: Environment) -> Self {
		Self {
			bitcoin_network: environment.bitcoin_network,
			ethereum_chain_id: environment.ethereum_chain_id,
			polkadot_genesis_hash: environment.polkadot_genesis_hash,
		}
	}
}

#[rpc(server, client, namespace = "cf")]
/// The custom RPC endpoints for the state chain node.
pub trait CustomApi {
	/// Returns true if the current phase is the auction phase.
	#[method(name = "is_auction_phase")]
	fn cf_is_auction_phase(&self, at: Option<state_chain_runtime::Hash>) -> RpcResult<bool>;
	#[method(name = "eth_key_manager_address")]
	fn cf_eth_key_manager_address(
		&self,
		at: Option<state_chain_runtime::Hash>,
	) -> RpcResult<String>;
	#[method(name = "eth_state_chain_gateway_address")]
	fn cf_eth_state_chain_gateway_address(
		&self,
		at: Option<state_chain_runtime::Hash>,
	) -> RpcResult<String>;
	#[method(name = "cf_eth_asset")]
	fn cf_eth_asset(
		&self,
		at: Option<state_chain_runtime::Hash>,
		token_address: EthereumAddress,
	) -> RpcResult<Option<Asset>>;
	#[method(name = "eth_flip_token_address")]
	fn cf_eth_flip_token_address(&self, at: Option<state_chain_runtime::Hash>)
		-> RpcResult<String>;
	#[method(name = "eth_chain_id")]
	fn cf_eth_chain_id(&self, at: Option<state_chain_runtime::Hash>) -> RpcResult<u64>;
	/// Returns the eth vault in the form [agg_key, active_from_eth_block]
	#[method(name = "eth_vault")]
	fn cf_eth_vault(&self, at: Option<state_chain_runtime::Hash>) -> RpcResult<(String, u32)>;
	#[method(name = "tx_fee_multiplier")]
	fn cf_tx_fee_multiplier(&self, at: Option<state_chain_runtime::Hash>) -> RpcResult<u64>;
	// Returns the Auction params in the form [min_set_size, max_set_size]
	#[method(name = "auction_parameters")]
	fn cf_auction_parameters(&self, at: Option<state_chain_runtime::Hash>)
		-> RpcResult<(u32, u32)>;
	#[method(name = "min_funding")]
	fn cf_min_funding(&self, at: Option<state_chain_runtime::Hash>) -> RpcResult<NumberOrHex>;
	#[method(name = "current_epoch")]
	fn cf_current_epoch(&self, at: Option<state_chain_runtime::Hash>) -> RpcResult<u32>;
	#[method(name = "epoch_duration")]
	fn cf_epoch_duration(&self, at: Option<state_chain_runtime::Hash>) -> RpcResult<u32>;
	#[method(name = "current_epoch_started_at")]
	fn cf_current_epoch_started_at(&self, at: Option<state_chain_runtime::Hash>) -> RpcResult<u32>;
	#[method(name = "authority_emission_per_block")]
	fn cf_authority_emission_per_block(
		&self,
		at: Option<state_chain_runtime::Hash>,
	) -> RpcResult<NumberOrHex>;
	#[method(name = "backup_emission_per_block")]
	fn cf_backup_emission_per_block(
		&self,
		at: Option<state_chain_runtime::Hash>,
	) -> RpcResult<NumberOrHex>;
	#[method(name = "flip_supply")]
	fn cf_flip_supply(
		&self,
		at: Option<state_chain_runtime::Hash>,
	) -> RpcResult<(NumberOrHex, NumberOrHex)>;
	#[method(name = "accounts")]
	fn cf_accounts(
		&self,
		at: Option<state_chain_runtime::Hash>,
	) -> RpcResult<Vec<(AccountId32, String)>>;
	#[method(name = "account_info")]
	fn cf_account_info(
		&self,
		account_id: AccountId32,
		at: Option<state_chain_runtime::Hash>,
	) -> RpcResult<RpcAccountInfo>;
	#[method(name = "account_info_v2")]
	fn cf_account_info_v2(
		&self,
		account_id: AccountId32,
		at: Option<state_chain_runtime::Hash>,
	) -> RpcResult<RpcAccountInfoV2>;
	#[method(name = "penalties")]
	fn cf_penalties(
		&self,
		at: Option<state_chain_runtime::Hash>,
	) -> RpcResult<Vec<(Offence, RpcPenalty)>>;
	#[method(name = "suspensions")]
	fn cf_suspensions(&self, at: Option<state_chain_runtime::Hash>) -> RpcResult<RpcSuspensions>;
	#[method(name = "generate_gov_key_call_hash")]
	fn cf_generate_gov_key_call_hash(
		&self,
		call: Vec<u8>,
		at: Option<state_chain_runtime::Hash>,
	) -> RpcResult<GovCallHash>;
	#[method(name = "auction_state")]
	fn cf_auction_state(&self, at: Option<state_chain_runtime::Hash>)
		-> RpcResult<RpcAuctionState>;
	#[method(name = "pool_sqrt_price")]
	fn cf_pool_sqrt_price(
		&self,
		from: Asset,
		to: Asset,
		at: Option<state_chain_runtime::Hash>,
	) -> RpcResult<Option<SqrtPriceQ64F96>>;
	#[method(name = "swap_rate")]
	fn cf_pool_swap_rate(
		&self,
		from: Asset,
		to: Asset,
		amount: NumberOrHex,
		at: Option<state_chain_runtime::Hash>,
	) -> RpcResult<RpcSwapOutput>;
	#[method(name = "environment")]
	fn cf_environment(&self, at: Option<state_chain_runtime::Hash>) -> RpcResult<RpcEnvironment>;
}

/// An RPC extension for the state chain node.
pub struct CustomRpc<C, B> {
	pub client: Arc<C>,
	pub _phantom: PhantomData<B>,
}

impl<C, B> CustomRpc<C, B>
where
	B: sp_runtime::traits::Block<Hash = state_chain_runtime::Hash>,
	C: sp_api::ProvideRuntimeApi<B> + Send + Sync + 'static + HeaderBackend<B>,
	C::Api: CustomRuntimeApi<B>,
{
	fn query_block_id(&self, from_rpc: Option<<B as BlockT>::Hash>) -> sp_api::BlockId<B> {
		sp_api::BlockId::hash(from_rpc.unwrap_or_else(|| self.client.info().best_hash))
	}
}

fn to_rpc_error<E: std::error::Error + Send + Sync + 'static>(e: E) -> jsonrpsee::core::Error {
	CallError::from_std_error(e).into()
}

impl<C, B> CustomApiServer for CustomRpc<C, B>
where
	B: sp_runtime::traits::Block<Hash = state_chain_runtime::Hash>,
	C: sp_api::ProvideRuntimeApi<B> + Send + Sync + 'static + HeaderBackend<B>,
	C::Api: CustomRuntimeApi<B>,
{
	fn cf_is_auction_phase(&self, at: Option<<B as BlockT>::Hash>) -> RpcResult<bool> {
		self.client
			.runtime_api()
			.cf_is_auction_phase(&self.query_block_id(at))
			.map_err(to_rpc_error)
	}
	fn cf_eth_flip_token_address(&self, at: Option<<B as BlockT>::Hash>) -> RpcResult<String> {
		self.client
			.runtime_api()
			.cf_eth_flip_token_address(&self.query_block_id(at))
			.map_err(to_rpc_error)
			.map(hex::encode)
	}
	fn cf_eth_asset(
		&self,
		at: Option<<B as BlockT>::Hash>,
		token_address: EthereumAddress,
	) -> RpcResult<Option<Asset>> {
		self.client
			.runtime_api()
			.cf_eth_asset(&self.query_block_id(at), token_address)
			.map_err(to_rpc_error)
	}
	fn cf_eth_state_chain_gateway_address(
		&self,
		at: Option<<B as BlockT>::Hash>,
	) -> RpcResult<String> {
		self.client
			.runtime_api()
			.cf_eth_state_chain_gateway_address(&self.query_block_id(at))
			.map_err(to_rpc_error)
			.map(hex::encode)
	}
	fn cf_eth_key_manager_address(&self, at: Option<<B as BlockT>::Hash>) -> RpcResult<String> {
		self.client
			.runtime_api()
			.cf_eth_key_manager_address(&self.query_block_id(at))
			.map_err(to_rpc_error)
			.map(hex::encode)
	}
	fn cf_eth_chain_id(&self, at: Option<<B as BlockT>::Hash>) -> RpcResult<u64> {
		self.client
			.runtime_api()
			.cf_eth_chain_id(&self.query_block_id(at))
			.map_err(to_rpc_error)
	}
	fn cf_eth_vault(&self, at: Option<<B as BlockT>::Hash>) -> RpcResult<(String, u32)> {
		self.client
			.runtime_api()
			.cf_eth_vault(&self.query_block_id(at))
			.map(|(public_key, active_from_block)| (hex::encode(public_key), active_from_block))
			.map_err(to_rpc_error)
	}
	// FIXME: Respect the block hash argument here
	fn cf_tx_fee_multiplier(&self, _at: Option<<B as BlockT>::Hash>) -> RpcResult<u64> {
		Ok(TX_FEE_MULTIPLIER as u64)
	}
	fn cf_auction_parameters(&self, at: Option<<B as BlockT>::Hash>) -> RpcResult<(u32, u32)> {
		self.client
			.runtime_api()
			.cf_auction_parameters(&self.query_block_id(at))
			.map_err(to_rpc_error)
	}
	fn cf_min_funding(&self, at: Option<<B as BlockT>::Hash>) -> RpcResult<NumberOrHex> {
		self.client
			.runtime_api()
			.cf_min_funding(&self.query_block_id(at))
			.map_err(to_rpc_error)
			.map(Into::into)
	}
	fn cf_current_epoch(&self, at: Option<<B as BlockT>::Hash>) -> RpcResult<u32> {
		self.client
			.runtime_api()
			.cf_current_epoch(&self.query_block_id(at))
			.map_err(to_rpc_error)
	}
	fn cf_epoch_duration(&self, at: Option<<B as BlockT>::Hash>) -> RpcResult<u32> {
		self.client
			.runtime_api()
			.cf_epoch_duration(&self.query_block_id(at))
			.map_err(to_rpc_error)
	}
	fn cf_current_epoch_started_at(&self, at: Option<<B as BlockT>::Hash>) -> RpcResult<u32> {
		self.client
			.runtime_api()
			.cf_current_epoch_started_at(&self.query_block_id(at))
			.map_err(to_rpc_error)
	}
	fn cf_authority_emission_per_block(
		&self,
		at: Option<<B as BlockT>::Hash>,
	) -> RpcResult<NumberOrHex> {
		self.client
			.runtime_api()
			.cf_authority_emission_per_block(&self.query_block_id(at))
			.map_err(to_rpc_error)
			.map(Into::into)
	}
	fn cf_backup_emission_per_block(
		&self,
		at: Option<<B as BlockT>::Hash>,
	) -> RpcResult<NumberOrHex> {
		self.client
			.runtime_api()
			.cf_backup_emission_per_block(&self.query_block_id(at))
			.map_err(to_rpc_error)
			.map(Into::into)
	}
	fn cf_flip_supply(
		&self,
		at: Option<<B as BlockT>::Hash>,
	) -> RpcResult<(NumberOrHex, NumberOrHex)> {
		self.client
			.runtime_api()
			.cf_flip_supply(&self.query_block_id(at))
			.map_err(to_rpc_error)
			.map(|(issuance, offchain)| (issuance.into(), offchain.into()))
	}
	fn cf_accounts(
		&self,
		at: Option<<B as BlockT>::Hash>,
	) -> RpcResult<Vec<(AccountId32, String)>> {
		Ok(self
			.client
			.runtime_api()
			.cf_accounts(&self.query_block_id(at))
			.map_err(to_rpc_error)?
			.into_iter()
			.map(|(account_id, vanity_name_bytes)| {
				// we can use from_utf8_lossy here because we're guaranteed utf8 when we
				// save the vanity name on the chain
				(account_id, String::from_utf8_lossy(&vanity_name_bytes).into_owned())
			})
			.collect())
	}
	fn cf_account_info(
		&self,
		account_id: AccountId32,
		at: Option<<B as BlockT>::Hash>,
	) -> RpcResult<RpcAccountInfo> {
		let account_info = self
			.client
			.runtime_api()
			.cf_account_info(&self.query_block_id(at), account_id)
			.map_err(to_rpc_error)?;

		Ok(RpcAccountInfo {
			balance: account_info.balance.into(),
			bond: account_info.bond.into(),
			last_heartbeat: account_info.last_heartbeat,
			is_live: account_info.is_live,
			is_activated: account_info.is_activated,
			online_credits: account_info.online_credits,
			reputation_points: account_info.reputation_points,
			state: account_info.state,
		})
	}
	fn cf_account_info_v2(
		&self,
		account_id: AccountId32,
		at: Option<<B as BlockT>::Hash>,
	) -> RpcResult<RpcAccountInfoV2> {
		let account_info = self
			.client
			.runtime_api()
			.cf_account_info_v2(&self.query_block_id(at), account_id)
			.map_err(to_rpc_error)?;

		Ok(RpcAccountInfoV2 {
			balance: account_info.balance.into(),
			bond: account_info.bond.into(),
			last_heartbeat: account_info.last_heartbeat,
			online_credits: account_info.online_credits,
			reputation_points: account_info.reputation_points,
			keyholder_epochs: account_info.keyholder_epochs,
			is_current_authority: account_info.is_current_authority,
			is_current_backup: account_info.is_current_backup,
			is_qualified: account_info.is_qualified,
			is_online: account_info.is_online,
			is_bidding: account_info.is_bidding,
			bound_redeem_address: account_info.bound_redeem_address,
		})
	}
	fn cf_penalties(
		&self,
		at: Option<<B as BlockT>::Hash>,
	) -> RpcResult<Vec<(Offence, RpcPenalty)>> {
		Ok(self
			.client
			.runtime_api()
			.cf_penalties(&self.query_block_id(at))
			.map_err(to_rpc_error)?
			.iter()
			.map(|(offence, runtime_api_penalty)| {
				(
					*offence,
					RpcPenalty {
						reputation_points: runtime_api_penalty.reputation_points,
						suspension_duration_blocks: runtime_api_penalty.suspension_duration_blocks,
					},
				)
			})
			.collect())
	}
	fn cf_suspensions(&self, at: Option<<B as BlockT>::Hash>) -> RpcResult<RpcSuspensions> {
		self.client
			.runtime_api()
			.cf_suspensions(&self.query_block_id(at))
			.map_err(to_rpc_error)
	}

	fn cf_generate_gov_key_call_hash(
		&self,
		call: Vec<u8>,
		at: Option<<B as BlockT>::Hash>,
	) -> RpcResult<GovCallHash> {
		self.client
			.runtime_api()
			.cf_generate_gov_key_call_hash(&self.query_block_id(at), call)
			.map_err(to_rpc_error)
	}

	fn cf_auction_state(&self, at: Option<<B as BlockT>::Hash>) -> RpcResult<RpcAuctionState> {
		let auction_state = self
			.client
			.runtime_api()
			.cf_auction_state(&self.query_block_id(at))
			.map_err(to_rpc_error)?;

		Ok(RpcAuctionState {
			blocks_per_epoch: auction_state.blocks_per_epoch,
			current_epoch_started_at: auction_state.current_epoch_started_at,
			redemption_period_as_percentage: auction_state.redemption_period_as_percentage,
			min_funding: auction_state.min_funding.into(),
			auction_size_range: auction_state.auction_size_range,
		})
	}

	fn cf_pool_sqrt_price(
		&self,
		from: Asset,
		to: Asset,
		at: Option<state_chain_runtime::Hash>,
	) -> RpcResult<Option<SqrtPriceQ64F96>> {
		self.client
			.runtime_api()
			.cf_pool_sqrt_price(&self.query_block_id(at), from, to)
			.map_err(to_rpc_error)
	}

	fn cf_pool_swap_rate(
		&self,
		from: Asset,
		to: Asset,
		amount: NumberOrHex,
		at: Option<state_chain_runtime::Hash>,
	) -> RpcResult<RpcSwapOutput> {
		self.client
			.runtime_api()
			.cf_pool_simulate_swap(
				&self.query_block_id(at),
				from,
				to,
				cf_utilities::try_parse_number_or_hex(amount).and_then(|amount| {
					if amount == 0 {
						Err(anyhow::anyhow!("Swap input amount cannot be zero."))
					} else {
						Ok(amount)
					}
				})?,
			)
			.map_err(to_rpc_error)
			.and_then(|r| {
				r.ok_or(jsonrpsee::core::Error::from(anyhow::anyhow!("Unable to process swap.")))
			})
			.map(RpcSwapOutput::from)
	}

	fn cf_environment(&self, at: Option<state_chain_runtime::Hash>) -> RpcResult<RpcEnvironment> {
		self.client
			.runtime_api()
			.cf_environment(&self.query_block_id(at))
			.map_err(to_rpc_error)
			.map(RpcEnvironment::from)
	}
}
