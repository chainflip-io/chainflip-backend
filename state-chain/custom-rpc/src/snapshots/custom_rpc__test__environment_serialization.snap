---
source: state-chain/custom-rpc/src/lib.rs
assertion_line: 1690
expression: "serde_json::to_value(env).unwrap()"
---
{"funding":{"minimum_funding_amount":0,"redemption_tax":0},"ingress_egress":{"channel_opening_fees":{"Arbitrum":1000,"Bitcoin":0,"Ethereum":1000,"Polkadot":1000},"egress_dust_limits":{"Arbitrum":{"ETH":0,"USDC":"0xffffffffffffffff"},"Bitcoin":{"BTC":0},"Ethereum":{"ETH":0,"FLIP":"0xffffffffffffffffffffffffffffffff","USDC":"0x7ffffffffffffffe","USDT":0},"Polkadot":{"DOT":0}},"egress_fees":{"Arbitrum":{"ETH":0,"USDC":null},"Bitcoin":{"BTC":0},"Ethereum":{"ETH":0,"FLIP":"0xffffffffffffffffffffffffffffffff","USDC":null,"USDT":null},"Polkadot":{"DOT":"0x7ffffffffffffffe"}},"ingress_fees":{"Arbitrum":{"ETH":0,"USDC":null},"Bitcoin":{"BTC":0},"Ethereum":{"ETH":0,"FLIP":"0xffffffffffffffffffffffffffffffff","USDC":null,"USDT":null},"Polkadot":{"DOT":"0x7ffffffffffffffe"}},"minimum_deposit_amounts":{"Arbitrum":{"ETH":0,"USDC":"0xffffffffffffffff"},"Bitcoin":{"BTC":0},"Ethereum":{"ETH":0,"FLIP":"0xffffffffffffffff","USDC":"0x7ffffffffffffffe","USDT":0},"Polkadot":{"DOT":0}},"witness_safety_margins":{"Arbitrum":null,"Bitcoin":3,"Ethereum":3,"Polkadot":null}},"pools":{"fees":{"Arbitrum":{"ETH":{"limit_order_fee_hundredth_pips":0,"limit_order_total_fees_earned":{"base":"0x0","quote":"0x0"},"limit_total_swap_inputs":{"base":"0x0","quote":"0x0"},"quote_asset":{"asset":"USDC","chain":"Ethereum"},"range_order_fee_hundredth_pips":100,"range_order_total_fees_earned":{"base":"0x0","quote":"0x0"},"range_total_swap_inputs":{"base":"0x0","quote":"0x0"}},"USDC":{"limit_order_fee_hundredth_pips":0,"limit_order_total_fees_earned":{"base":"0x0","quote":"0x0"},"limit_total_swap_inputs":{"base":"0x0","quote":"0x0"},"quote_asset":{"asset":"USDC","chain":"Ethereum"},"range_order_fee_hundredth_pips":100,"range_order_total_fees_earned":{"base":"0x0","quote":"0x0"},"range_total_swap_inputs":{"base":"0x0","quote":"0x0"}}},"Bitcoin":{"BTC":{"limit_order_fee_hundredth_pips":0,"limit_order_total_fees_earned":{"base":"0x0","quote":"0x0"},"limit_total_swap_inputs":{"base":"0x0","quote":"0x0"},"quote_asset":{"asset":"USDC","chain":"Ethereum"},"range_order_fee_hundredth_pips":100,"range_order_total_fees_earned":{"base":"0x0","quote":"0x0"},"range_total_swap_inputs":{"base":"0x0","quote":"0x0"}}},"Ethereum":{"ETH":null,"FLIP":{"limit_order_fee_hundredth_pips":0,"limit_order_total_fees_earned":{"base":"0x0","quote":"0x0"},"limit_total_swap_inputs":{"base":"0x0","quote":"0x0"},"quote_asset":{"asset":"USDC","chain":"Ethereum"},"range_order_fee_hundredth_pips":100,"range_order_total_fees_earned":{"base":"0x0","quote":"0x0"},"range_total_swap_inputs":{"base":"0x0","quote":"0x0"}},"USDC":null,"USDT":{"limit_order_fee_hundredth_pips":0,"limit_order_total_fees_earned":{"base":"0x0","quote":"0x0"},"limit_total_swap_inputs":{"base":"0x0","quote":"0x0"},"quote_asset":{"asset":"USDC","chain":"Ethereum"},"range_order_fee_hundredth_pips":100,"range_order_total_fees_earned":{"base":"0x0","quote":"0x0"},"range_total_swap_inputs":{"base":"0x0","quote":"0x0"}}},"Polkadot":{"DOT":{"limit_order_fee_hundredth_pips":0,"limit_order_total_fees_earned":{"base":"0x0","quote":"0x0"},"limit_total_swap_inputs":{"base":"0x0","quote":"0x0"},"quote_asset":{"asset":"USDC","chain":"Ethereum"},"range_order_fee_hundredth_pips":100,"range_order_total_fees_earned":{"base":"0x0","quote":"0x0"},"range_total_swap_inputs":{"base":"0x0","quote":"0x0"}}}}},"swapping":{"maximum_swap_amounts":{"Arbitrum":{"ETH":null,"USDC":0},"Bitcoin":{"BTC":0},"Ethereum":{"ETH":0,"FLIP":null,"USDC":"0x7ffffffffffffffe","USDT":null},"Polkadot":{"DOT":null}},"network_fee_hundredth_pips":1000000}}
